### PR TITLE
[view only] Block uploading: statemachine

### PR DESCRIPTION
**Pull request by folex**

### CODE REVIEW

File: statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  /**
+   * Move list of current DropPeer events from ControlSignals to call-site
+   * dropPeersRef is emptied on resource's acquisition, and filled with Nil after resource is used
+   * Using Resource this way guarantees exclusive access to data
+   *
+   * @return Resource with List of DropPeer signals
+   */
+  val dropPeers: Resource[F, Set[DropPeer]] =
+    Resource.make(dropPeersRef.tryTake.map(_.getOrElse(Set.empty)))(_ => dropPeersRef.tryPut(Set.empty).void)
```
Comment (DieMyst) [inline statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala]: are you sure that `tryTake` and `tryPut` is ok here?

File: statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+   * Retrieves block receipt, async blocks until there's a receipt
+   */
+  val receipt: F[BlockReceipt] = receiptRef.take
+
+  /**
+   * Stores vm hash to memory, so node can retrieve it for block manifest uploading
+   */
+  def putVmHash(hash: ByteVector): F[Unit] = hashRef.put(hash)
+
+  def setVmHash(hash: ByteVector): F[Unit] = hashRef.tryTake >> hashRef.put(hash)
```
Comment (DieMyst) [inline statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala]: `tryTake` = `take` i think

File: statemachine/src/main/scala/com/github/jtendermint/jabci/socket/MyTSocket.java [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+import com.google.protobuf.CodedInputStream;
+import com.google.protobuf.GeneratedMessageV3;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+/**
+ * The TSocket acts as the main socket connection to the Tendermint-Node
+ *
+ * @author srmo, wolfposd
+ */
+public class MyTSocket extends ASocket {
```
Comment (DieMyst) [inline statemachine/src/main/scala/com/github/jtendermint/jabci/socket/MyTSocket.java]: rename it?

File: statemachine/src/main/scala/com/github/jtendermint/jabci/socket/MyTSocket.java [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+    public static final int DEFAULT_LISTEN_SOCKET_PORT = 26658;
+    public static final String INFO_SOCKET = "-Info";
+    public static final String MEMPOOL_SOCKET = "-MemPool";
+    public static final String CONSENSUS_SOCKET = "-Consensus";
+    public static final int DEFAULT_LISTEN_SOCKET_TIMEOUT = 1000;
+
+//    private static final Logger System.out.format()TSocket: .getLogger(MyTSocket.class + "\n);
```
Comment (DieMyst) [inline statemachine/src/main/scala/com/github/jtendermint/jabci/socket/MyTSocket.java]: comments

File: statemachine/src/main/scala/com/github/jtendermint/jabci/socket/MyTSocket.java [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    }
+
+    /**
+     * Start listening on the specified port
+     *
+     * @param portNumber
+     * @param socketTimeout
+     */
+    public void start(final int portNumber, final int socketTimeout) {
+        System.out.format("TSocket: starting serversocket" + "\n");
```
Comment (DieMyst) [inline statemachine/src/main/scala/com/github/jtendermint/jabci/socket/MyTSocket.java]: logs somehow?

File: statemachine/src/main/scala/fluence/statemachine/AbciService.scala [hunk 6]
```diff
@@ -39,25 +46,63 @@ import scala.language.higherKinds
...
+  tendermintRpc: TendermintHttpRpc[F]
+)(implicit hasher: Hasher[ByteVector, ByteVector], log: Log[F]) {
 
   import AbciService._
 
+  private def checkBlock(height: Long)(implicit log: Log[F]): Unit =
+    (for {
+      block ← tendermintRpc
+        .block(height)
+        .leftSemiflatMap(e ⇒ Log[F].warn(s"RPC Block[$height] failed", e))
```
Comment (DieMyst) [inline statemachine/src/main/scala/fluence/statemachine/AbciService.scala]: is it ok to use logs as result in left projection?

File: statemachine/src/main/scala/fluence/statemachine/AbciService.scala [hunk 7]
```diff
@@ -73,21 +118,64 @@ class AbciService[F[_]: Monad](
...
+            log.error(s"Got wrong receipt height. height: $blockHeight, receipt: ${b.receipt.height}")
+      )
+
+      // Check previous block for correctness, for debugging purposes
+      _ = if (blockHeight > 1 && receipt.forall(_.`type` == ReceiptType.New)) checkBlock(blockHeight - 1)
+
+      // Do not use receipt in app hash if there's no txs in a block, so empty blocks have the same appHash as
+      // previous non-empty ones. This is because Tendermint stops producing empty blocks only after
+      // at least 2 blocks have the same appHash. Otherwise, empty blocks would be produced indefinitely.
+      // TODO: use appHash for the previous block instead of vmHash.pure[F]
```
Comment (DieMyst) [inline statemachine/src/main/scala/fluence/statemachine/AbciService.scala]: don't this todo blocks anything? or will break backward compatibility. Maybe it is better to do this task in this PR?

File: statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala (hunk 1 above)
Comment (folex) [reply statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala]: yep

Comment (folex) [reply statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala]: it's pretty old code btw :)

File: statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala (hunk 2 above)
Comment (folex) [reply statemachine/control/src/main/scala/fluence/statemachine/control/ControlSignalsImpl.scala]: Nope, the reason for tryTake here was no clear MVar iff there was something there. But that code is obsolete now anyway.

File: statemachine/src/main/scala/com/github/jtendermint/jabci/socket/MyTSocket.java (hunk 3 above)
Comment (folex) [reply statemachine/src/main/scala/com/github/jtendermint/jabci/socket/MyTSocket.java]: remove it!

File: statemachine/src/main/scala/fluence/statemachine/AbciService.scala (hunk 6 above)
Comment (folex) [reply statemachine/src/main/scala/fluence/statemachine/AbciService.scala]: why not?

**Pull request closed without merging**